src/sleep/sleep.c

```diff
@@ -263,40 +263,95 @@ static int execute(
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
+                previous_discharge_time = get_battery_discharge_rate();
+                if (previous_discharge_time <= 0)
+                        log_error_errno(errno, "Error fetching battery discharge rate: %m");
+                else
+                        suspend_interval = ((previous_discharge_time * 60) - 30) * USEC_PER_MINUTE;
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s",
+                        FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_MONOTONIC);
+
+                        current_capacity = read_battery_capacity();
+                        if (current_capacity < 0)
+                                return log_error_errno(errno, "Error fetching battery capacity: %m");
 
-        tfd = safe_close(tfd);
+                        log_debug("%d%% battery discharged in %s time",
+                                        last_capacity - current_capacity,
+                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, USEC_PER_HOUR));
+
+                        estimated_discharge_rate = (last_capacity - current_capacity) / (after_timestamp - before_timestamp);
+
+                        log_debug("Battery will discharge in %s time", FORMAT_TIMESPAN(estimated_discharge_rate, USEC_PER_HOUR));
+
+                        if (estimated_discharge_rate > 0 && estimated_discharge_rate < 1000) {
+                                r = put_battery_discharge_rate(estimated_discharge_rate);
+                                if (r < 0)
+                                        return log_warning_errno(r, "Failed to update battery discharge rate: ignoring %m");
+                        }
+                        return 0;
+                }
+
+                tfd = safe_close(tfd);
+
+                log_debug("Attempting to estimate battery after waking from %s timer",
```

Comment:
this needs the unit, not just 'timer' - `after waking from %s hours sleep timer`